pkg/certificates/bootstrap/bootstrap.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bootstrap
+
+import (
+	"crypto/x509/pkix"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	certificates "k8s.io/api/certificates/v1beta1"
+	v12 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v13 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	certificatesclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/certificate"
+	"k8s.io/client-go/util/certificate/csr"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+)
+
+const (
+	keyBytesValue = "key-bytes"
+)
+
+// LoadClientCertForService requests a client cert for a component if the the certDir does not contain a certificate.
+// The certificate and key file are stored in certDir.
+func LoadClientCertForNode(bootstrapClient certificatesclient.CertificatesV1beta1Interface, certStore certificate.Store, name types.NodeName) error {
+
+	var keyData []byte
+	if cert, err := certStore.Current(); err == nil {
+		if cert.PrivateKey != nil {
+			keyData, err = certutil.MarshalPrivateKeyToPEM(cert.PrivateKey)
+			if err != nil {
+				keyData = nil
+			}
+		}
+	}
+	if !verifyKeyData(keyData) {
+		var err error
+		glog.V(2).Infof("No valid private key found for bootstrapping, creating a new one")
+		keyData, err = certutil.MakeEllipticPrivateKeyPEM()
+		if err != nil {
+			return err
+		}
+	}
+
+	certData, err := RequestKubeVirtCertificate(bootstrapClient.CertificateSigningRequests(), keyData, "kubevirt.io:system:nodes:"+string(name))
+	if err != nil {
+		return err
+	}
+	if _, err := certStore.Update(certData, keyData); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// LoadClientCertForService fetches a client key from a secret or generates one and updates the secret.
+// Afterwards it creates as certificate signing request and fetches the result after it got signed.
+func LoadClientCertForService(client kubecli.KubevirtClient, certStore certificate.Store, serviceName string, namespace string) error {
+
+	keyData, err := LoadKeyFromSecret(client, namespace, "kubevirt.io-system-certificates-"+serviceName)
```

Comment:
I think that storing the certificate in the secret is nonsens actually. Will remove that. Every pod can have it's own cert for cluster components. It will then have the `podIP` and in the case of `virt-api` the service name included in the certificate.